### PR TITLE
Wafer Map  - codebase refactor & zoom handling

### PR DESCRIPTION
**Pull request by DStavilaNI**
# Pull Request

## 🤨 Rationale

This PR implements an event handling architecture complete with a Zoom handler. The PR also performs an extensive refactor of the already existing code base -> index file and types

The PR contains "TODO" items and references to a [HoverHandler ](https://dev.azure.com/ni/DevCentral/_boards/board/t/OptimalPlus%20Dokker/Work%20Items/?workitem=2211981)module - this module exists, however it was removed from the current Wafermap version as it is still a work-in-progress and does not meet yet the required coding/usability standards. Rather than submitting partial code, we decided to do so in a future PR. This is expected to happen towards the end of this week. 

## 👩‍💻 Implementation

- In this PR, a lot of changes were applied to the existing Wafer Map code-base to improve code readability, modularity and ease of maintenance (index.ts, types.ts). 
- A top level EventHandler was added
- A ZoomHandler was added

## 🧪 Testing

- Manual testing has been done. There are no changes to existing functionality.

## ✅ Checklist

- [x] I have updated the project documentation to reflect my changes or determined no changes are needed.

### CODE REVIEW

File: change/@ni-nimble-components-5067d35f-244c-42ff-b498-4329212f03ce.json
Comment (jattasNI) [inline change/@ni-nimble-components-5067d35f-244c-42ff-b498-4329212f03ce.json:1]: You can approve the Chromatic diffs even though they're unrelated to your change. They are known intermittencies.

File: packages/nimble-components/src/wafer-map/modules/event-handler.ts
Comment (jattasNI) [inline packages/nimble-components/src/wafer-map/modules/event-handler.ts:8]: What do you think of alternate names like `UserInteractionHandler` or `InteractionHandler`? "Event" is a pretty generic term which could refer to events that are from various sources so it would be nice to be more specific.

If you change this then we should update the corresponding type names too.

File: packages/nimble-components/src/wafer-map/index.ts
Comment (jattasNI) [inline packages/nimble-components/src/wafer-map/index.ts:79]: It's worth pausing to discuss this new public element API. Some questions/thoughts:
1. naming: what does the "last" part of the name mean? I'd expect to be able to access the current selection (or undefined if nothing is selected) but "last" to me implies something happened in the past. How about just `selectedDie`?
2. API type: is a property getter of type `WaferMapDie` appropriate? This can't be accessed as an HTML attribute and doesn't provide any notification when selection changes so frameworks like Angular can't data bind to it. If we wanted to make it an attribute with a changed event we could change its type to be a number that indexes into the `dies` array.
3. it'd be good to update these places to document this API:
   - API section of the spec
   - storybook for the wafer map
   - if documenting it inline, use a JSDoc style coment `/**` and provide more detail (or just delete the comment above)

File: packages/nimble-components/src/wafer-map/index.ts
Comment (jattasNI) [inline packages/nimble-components/src/wafer-map/index.ts:224]: I think this data transformation should be the responsibility of the various handlers. Putting it in the wafer map means that the concepts like zoom handler and hover handler leak out to this level but it would be better to encapsulate those. The handlers could hopefully take a reference to the top level wafer map and use it to look up whatever they need and transform it to the form they need.

File: packages/nimble-components/src/wafer-map/types.ts
Comment (jattasNI) [inline packages/nimble-components/src/wafer-map/types.ts:54]: I'm open to other opinions on this, but I think it would be easier to follow if each of these interfaces moved to the handler files that use them. The `types.ts` file is more useful for types that are expected to be used publicly.

File: packages/nimble-components/src/wafer-map/template.ts
Comment (jattasNI) [inline packages/nimble-components/src/wafer-map/template.ts:27]: Does this new layer (and associated CSS) need to be part of this PR or can it wait for the hover PR (possibly with a TODO instead if that is helpful for tracking)?

File: packages/nimble-components/src/wafer-map/modules/zoom-handler.ts
Comment (jattasNI) [inline packages/nimble-components/src/wafer-map/modules/zoom-handler.ts:63]: nit: `reScale` looks funny to me; I'd vote for `rescale`

File: packages/nimble-components/src/wafer-map/modules/zoom-handler.ts
Comment (jattasNI) [inline packages/nimble-components/src/wafer-map/modules/zoom-handler.ts:25]: Why is `renderingFunction` a variable instead of a method? It doesn't look like it's passed around anywhere by reference. Making it a method would be easier to follow and allow removing the `this.renderingFunction === undefined` check below.

File: packages/nimble-components/src/wafer-map/modules/zoom-handler.ts
Comment (jattasNI) [inline packages/nimble-components/src/wafer-map/modules/zoom-handler.ts:15]: It would be helpful to document the pattern we're using here and describe what this does. Something like this?
```suggestion
    /*
     * This "event" is triggered when the user zooms the wafer map.
     * For simplicity this is just a callback function so it can only have one subscriber right now.
     */
    public onZoom: ((event: ZoomEvent) => void) | undefined;
```

File: packages/nimble-components/src/wafer-map/modules/event-handler.ts
Comment (jattasNI) [inline packages/nimble-components/src/wafer-map/modules/event-handler.ts:29]: Instead of exposing `rescale` publicly, I think it would be cleaner for the zoom handler to expose `onBeforeZoom` and `onAfterZoom` "events". Then the zoom handler could call `rescale` itself in between when those events are fired. That keeps the logic of zooming encapsulated within the zoom handler and limits its API to exposing more predictable lifecycle events.

File: packages/nimble-components/src/wafer-map/index.ts
Comment (jattasNI) [reply packages/nimble-components/src/wafer-map/index.ts:79]: We could also remove this for now and have this discussion on a future PR, since it looks like this is just staging work for the hover feature.

File: packages/nimble-components/src/wafer-map/modules/event-handler.ts
Comment (jattasNI) [inline packages/nimble-components/src/wafer-map/modules/event-handler.ts:19]: Instead of a public API to attach events, can this work happen in the constructor?

File: change/@ni-nimble-components-5067d35f-244c-42ff-b498-4329212f03ce.json
Comment (DStavilaNI) [reply change/@ni-nimble-components-5067d35f-244c-42ff-b498-4329212f03ce.json]: applied

File: change/@ni-nimble-components-5067d35f-244c-42ff-b498-4329212f03ce.json
Comment (DStavilaNI) [reply change/@ni-nimble-components-5067d35f-244c-42ff-b498-4329212f03ce.json:1]: Yup, noticed ... thanks for the go ahead :)

File: packages/nimble-components/src/wafer-map/modules/event-handler.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/modules/event-handler.ts:8]: I see your point, however I would still prefer to keep it as it is, since we might have indeed the need, in the future, to handle events coming from other sources as well, not just user interactions...

File: packages/nimble-components/src/wafer-map/modules/zoom-handler.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/modules/zoom-handler.ts]: applied

File: packages/nimble-components/src/wafer-map/modules/zoom-handler.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/modules/zoom-handler.ts:25]: I agree with removing the check - it was indeed some leftover code left from a previous iteration of the code (good catch here). It is no longer necessary. However, the `renderingFunction` is a property since it is expected to to populated by the corresponding method from the `renderModule` ... Of course, this could be wrapped in a function, but I did not see yet a strong case for it....

File: packages/nimble-components/src/wafer-map/modules/zoom-handler.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/modules/zoom-handler.ts:15]: applied

**Comment by jattasNI** [comment]
Friendly reminder that GitHub's silly UI is hiding some of my comments that still need to be considered:

https://user-images.githubusercontent.com/10500124/215794195-2f4b21cb-f940-4bc5-96c6-487c22d89de1.mov

File: packages/nimble-components/src/wafer-map/modules/event-handler.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/modules/event-handler.ts:19]: I think that it could and I did consider this, it's just that in this case we would need to modify the constructor to get also a reference to the `wafermap`, or to include it in the `EventHandlerData` ... Either way, by having it with this granularity level, it felt like the code was a little bit easier to follow ... I did not see a big enough gain in doing so at the moment.

File: packages/nimble-components/src/wafer-map/index.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/index.ts:223]: I understand what you mean, and I did consider this approach during the refactor process, however the issue with this is that a lot of the required properties are set as private in the current implementation of the Wafermap, and such would not be accessible from within the `EventHandler `. I would definitely keep this as a potential future improvement idea, but I am not sure if it makes sense to do this kind of a change now, especially since these 4 elements (WaferMap, ZoomHandler, EventHandler and later the HoverHandler) are and will stay very much coupled ...

File: packages/nimble-components/src/wafer-map/types.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/types.ts:54]: Tbh ... I also do not have a strong opinion on this. I am open to this change. The only issue is the presence of the `HoverHandlerData`, which I do not have a place yet to relocate to for this refactor and would need to be left here until the next PR or removed and re-added later ... which I would prefer not to do :) - it would impact other interfaces...

Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/types.ts:54]: I did apply this change, however I did leave for now the `HoverHandlerData ` within the "types" folder - I think it would be best if this change is applied in the follow-up PR once we add the actual `HoverHandler `code...

File: packages/nimble-components/src/wafer-map/template.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/template.ts:27]: Yup, I think it can be removed until the next PR.

File: packages/nimble-components/src/wafer-map/index.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/index.ts:78]: I removed it from the index.ts file. However I left it inside the `EventHandler` and renamed to `selectedDie` - I do not see any reason why it couldn't be named this way. The code will be finalized in the following PR.

File: packages/nimble-components/src/wafer-map/index.ts
Comment (jattasNI) [reply packages/nimble-components/src/wafer-map/index.ts:222]: I was worried that would be the case. The best workaround I can think of would be to make those wafer map properties `public` but tag them with `@internal` comments to indicate that they're not part of the public API. We do this for other components in quite a few places.

I still think this would be a step in the right direction to shrink the API surface of the handlers by encapsulating more. But I admit it might be messy or we might need to see how it works as we add more handlers so I'm ok deferring this.

File: packages/nimble-components/src/wafer-map/modules/event-handler.ts
Comment (jattasNI) [reply packages/nimble-components/src/wafer-map/modules/event-handler.ts:32]: The advantage would be to simplify the public API of each of the "handler" classes. If you can pass everything you need in the constructor then it's one less API to worry about. It saves us from asking questions like "what if `attachEvents` is called twice?" or "why isn't there a `detachEvents`?" or "what if you forget to call `attachEvents`?"

I still think it would be an improvement. I believe it would be pretty simple to include the wafer map as part of `EventHandlerData`.

**Review by jattasNI — COMMENTED** [review]
Looking much better! I'll wait for one more review round to see the resolution of the open comment threads and get a clean build.

File: packages/nimble-components/src/wafer-map/index.ts
Comment (jattasNI) [reply packages/nimble-components/src/wafer-map/index.ts:78]: Maybe I'm not following or maybe you haven't pushed some local changes, but I still see the old API here.

File: packages/nimble-components/src/wafer-map/modules/event-handler.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/modules/event-handler.ts]: ok ... applied

File: packages/nimble-components/src/wafer-map/index.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/index.ts]: It's definitely not there anymore :)

File: packages/nimble-components/src/wafer-map/index.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/index.ts:216]: We can get back to this subject in subsequent PRs if it will be the case.

File: packages/nimble-components/src/wafer-map/index.ts
Comment (rajsite) [inline packages/nimble-components/src/wafer-map/index.ts:157]: Treat this name as a method that does some action instead of how it's used. It is not a property that should be reassigned with a new method like the name `on-` suggests. Also use the built-in [emit helper](https://github.com/ni/nimble/blob/0bb622feed161189d38bb34cd8761c8bc752cf03/packages/nimble-components/src/menu-button/index.ts#L131) 

```suggestion
    private function emitDieSelected(die: WaferMapDie): void {
        this.$emit('die-selected', { detail: { die } });
    };
```

And below the adapter would be:
```ts
        const eventCallbacks: EventCallbacks = {
            dieSelected: (die) => this.emitDieSelected(die)
        };
```

File: packages/nimble-components/src/wafer-map/index.ts
Comment (rajsite) [inline packages/nimble-components/src/wafer-map/index.ts:142]: It seems strange to me to need to destroy and recreate all the event handlers during each render. Once the correct attach and detach behavior is implemented in this PR we should make sure this is actually necessary to do every render.

**Review by rajsite — CHANGES_REQUESTED** [review]

File: packages/nimble-components/src/wafer-map/index.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/index.ts:75]: This field will be used within the HoverHandler  ... it's part of the object that will be passed to the constructor once the handler is initialized. As I mentioned previously, the Handler is not yet ready to be submitted and I removed the corresponding code, however I wanted to keep still the corresponding data structures.

File: packages/nimble-components/src/wafer-map/modules/event-handler.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/modules/event-handler.ts:49]: applied

File: packages/nimble-components/src/wafer-map/modules/event-handler.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/modules/event-handler.ts:43]: applied

File: packages/nimble-components/src/wafer-map/index.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/index.ts:142]: applied

File: packages/nimble-components/src/wafer-map/modules/zoom-handler.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/modules/zoom-handler.ts]: applied

File: packages/nimble-components/src/wafer-map/index.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/index.ts:168]: applied

File: packages/nimble-components/src/wafer-map/index.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/index.ts]: This is done mainly because of the current implementation of the DataManager - which has to be recreated not just when new data is being fed to the component, but also when the size of the wafermap is changing (canvasSideLength). This is happening because of the computations being performed within. Since this module is also used by the rendering module and EventCoordinator, both have to be recreated as well.

File: packages/nimble-components/src/wafer-map/modules/event-handler.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/modules/event-handler.ts]: I applied most of your feedback:
1. EventHandler was renamed to EventCoordinator.
2. Moved all of the event registrations within the EventCoordinator, except the resizeObserver which would require some more extensive refactoring - the resizeObserver is used to track the canvasSideLength, which in turn is used (among other things) to decide to re-render or not the canvas. Unless you think that there is no other way, I would prefer to keep the resizeObserver as it is

File: packages/nimble-components/src/wafer-map/index.ts
Comment (jattasNI) [inline packages/nimble-components/src/wafer-map/index.ts:165]: We should document this event by adding it to the storybook for the wafer map.

**Review by jattasNI — APPROVED** [review]

**Review by rajsite — CHANGES_REQUESTED** [review]

File: packages/nimble-components/src/wafer-map/index.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/index.ts:109]: applied

File: packages/nimble-components/src/wafer-map/modules/event-coordinator.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/modules/event-coordinator.ts]: applied

File: packages/nimble-components/src/wafer-map/index.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/index.ts:165]: applied - as an action

File: packages/nimble-components/src/wafer-map/index.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/index.ts]: Renamed

File: packages/nimble-components/src/wafer-map/modules/event-coordinator.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/modules/event-coordinator.ts:9]: applied

File: packages/nimble-components/src/wafer-map/modules/zoom-handler.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/modules/zoom-handler.ts]: applied

File: packages/nimble-components/src/wafer-map/modules/zoom-handler.ts
Comment (rajsite) [inline packages/nimble-components/src/wafer-map/modules/zoom-handler.ts:69]: I did not understand how you can add the zoom event listeners multiple times without issue but the following docs describe why: https://github.com/d3/d3-zoom/blob/v3.0.0/README.md#zoom_on
Can you please include a comment that describes it?

```suggestion
            // D3 will automatically remove existing handlers when adding new ones
            // See: https://github.com/d3/d3-zoom/blob/v3.0.0/README.md#zoom_on
            .on('zoom', (event: ZoomEvent) => {
```

File: packages/nimble-components/src/wafer-map/index.ts
Comment (rajsite) [inline packages/nimble-components/src/wafer-map/index.ts:145]: Do you know when we will address recreating the event listeners every time there is an update?
We should make sure to track that tech debt if we can't address it in this PR.

File: packages/nimble-components/src/wafer-map/modules/rendering.ts
Comment (rajsite) [inline packages/nimble-components/src/wafer-map/modules/rendering.ts:29]: What is the goal of this event? It will potentially fire many times. i,e, continuously while zooming. We should not add it until it's needed so we can see the context.

File: packages/nimble-components/src/wafer-map/modules/rendering.ts
Comment (rajsite) [inline packages/nimble-components/src/wafer-map/modules/rendering.ts:7]: This does not get used as an event target in this PR so we should not make that change here.

**Comment by rajsite** [comment]
Reset @jattasNI since there has been a major architectural shift in the PR

File: packages/nimble-components/src/wafer-map/modules/data-manager.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/modules/data-manager.ts:81]: This will be eventually used by the HoverHandler, and the intent was to leave it there as it is, instead of removing and re-adding it later in a different PR. 
In the interest of time I will remove it though from the current PR.

File: packages/nimble-components/src/wafer-map/modules/zoom-handler.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/modules/zoom-handler.ts:18]: applied

File: packages/nimble-components/src/wafer-map/modules/zoom-handler.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/modules/zoom-handler.ts:69]: applied

File: packages/nimble-components/src/wafer-map/index.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/index.ts:145]: I would prefer this to be a follow-up PR as there are already other PRs, that we did queue and which depend on this one going in... we can create a github issue for this (if you prefer it this way) or we can track it in our internal AzDO, as part of the this feature. We can set it as tech debt.

Comment (rajsite) [reply packages/nimble-components/src/wafer-map/index.ts:145]: Please add a link here to where you chose to track it

File: packages/nimble-components/src/wafer-map/index.ts
Comment (rajsite) [inline packages/nimble-components/src/wafer-map/index.ts:131]: This is only used inside this file and can be made private

File: packages/nimble-components/src/wafer-map/index.ts
Comment (rajsite) [inline packages/nimble-components/src/wafer-map/index.ts:122]: This is only used inside this file and can be made private

**Review by rajsite — APPROVED** [review]

File: packages/nimble-components/src/wafer-map/modules/zoom-handler.ts
Comment (jattasNI) [inline packages/nimble-components/src/wafer-map/modules/zoom-handler.ts:11]: The `ZoomEvent` interface might not need to be exported since it appears to only be used within this file.

**Review by jattasNI — APPROVED** [review]
Thanks for your patience with all the rounds of feedback. I think the architecture has a better foundation to build on for upcoming features.

File: packages/nimble-components/src/wafer-map/index.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/index.ts:131]: applied

File: packages/nimble-components/src/wafer-map/index.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/index.ts:145]: https://dev.azure.com/ni/DevCentral/_workitems/edit/2294259

File: packages/nimble-components/src/wafer-map/index.ts
Comment (DStavilaNI) [reply packages/nimble-components/src/wafer-map/index.ts:122]: It is being used also in the tests (`wafer-map.spec.ts`) by the jasmine `spy` to track the rendering process. For now, I think it would be best to keep it `public `but marked as `internal`.

**Pull request merged**
